scripts/internal/prune_size_model.py

```diff
@@ -0,0 +1,246 @@
+        self.DebugLog("Threshold: {0}, num_xgrams: {1}".format(cur_threshold, cur_num_xgrams))
+
+        if cur_num_xgrams >= target_lower_threshold and cur_num_xgrams <= target_upper_threshold:
+            return ('success', None)
+
+        backtrack = False
+        if cur_num_xgrams < target_lower_threshold: # we overshot
+            if cur_threshold == self.initial_threshold:
+                # overshot with initial threshold, we should die
+                return ('overshoot', None)
+
+            # remove cur_threshold from history
+            self.history.pop()
+
+            if prev_threshold == cur_threshold:
+                # remove prev_threshold from history
+                self.history.pop()
+            backtrack = True
+            self.AdjustModelForOvershoot()
+
+        cur_target_num_xgrams = self.GetIntermediateTargetNumXgrams()
+        next_threshold = self.GetNextThreshold(cur_target_num_xgrams)
+
+        self.history.append([next_threshold, 0])
+
+        if backtrack:
+            return ('backtrack', [next_threshold, len(history) - 1])
+
+        return ('continue', next_threshold)
+
+    def GetIntermediateTargetNumXgrams(self):
+        """
+        This function gives us a point to aim for the next iteration.
+        If we're already very close to the target, we aim directly at the
+        target; otherwise we aim for a point short of the target.
+        """
+
+        cur_num_xgrams = self.GetCurNumXgrams()
+
+        if cur_num_xgrams > 1.5 * self.target_num_xgrams:
+            # If we're more than 1.5 times the target, aim to go only
+            # halfway to the target [in log-space], but to decrease
+            # the num-xgrams by no more than a factor of 4.
+            change_factor = (float(self.target_num_xgrams) / cur_num_xgrams) ** 0.5
+            if change_factor < 0.25:
+                change_factor = 0.25
+            return cur_num_xgrams * change_factor
+        elif cur_num_xgrams > 1.15 * self.target_num_xgrams:
+            # If we're between 1.15 and 1.5 times the target, aim to go
+            # two thirds of the way to the target.
+            change_factor = (float(self.target_num_xgrams) / cur_num_xgrams) ** 0.666
+
+            return cur_num_xgrams * change_factor
+        else:
+            # if we're within 15% of the target, aim directly for the target.
+            return self.target_num_xgrams
+
+    def GetNextThreshold(self, cur_target_num_xgrams):
+        """
+        This function searches for the value of next_num_xgrams >= cur_num_xgrams
+        that gives us a value as close as possible to our current
+        target num-xgrams. And return thd corresponding next_threshold
+        """
+        cur_threshold = self.GetCurThreshold()
+        cur_num_xgrams = self.GetCurNumXgrams()
+
+        # we use a simple binary search here
+        right = 10 * cur_threshold
+        left = cur_threshold # we never decrease the threshold
```

Comment:
I'd make the 0.005 a local variable, and make it e.g., like 0.0001 times the current threshold.